lib/puppet/application/agent.rb

```diff
@@ -358,16 +358,12 @@ def run_command
   end
 
   def fingerprint
-    host = Puppet::SSL::Host.new
-    unless cert = host.certificate || host.certificate_request
-      $stderr.puts _("Fingerprint asked but no certificate nor certificate request have yet been issued")
-      exit(1)
-      return
-    end
-    unless digest = cert.digest(options[:digest].to_s)
-      raise ArgumentError, _("Could not get fingerprint for digest '%{digest}'") % { digest: options[:digest] }
-    end
-    puts digest.to_s
+    sm = Puppet::SSL::StateMachine.new(onetime: true)
+    ssl_context = sm.ensure_client_certificate
+    puts Puppet::SSL::Digest.new(options[:digest].to_s, ssl_context.client_cert.to_der).to_s
```

Comment:
If we don't have a client cert yet (and autosigning is disabled), then `ssl_context.client_cert` can be nil.

Reply:
actually nevermind, `ensure_client_certificate` only returns if we download a cert, or it exits (since `onetime: true`).